File: services/notifier/src/telemetry.js
const { NodeTracerProvider } = require('@opentelemetry/sdk-trace-node');
const { BatchSpanProcessor } = require('@opentelemetry/sdk-trace-base');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-grpc');
const { resourceFromAttributes } = require('@opentelemetry/resources');
const { ATTR_SERVICE_NAME } = require('@opentelemetry/semantic-conventions');
const { AsyncHooksContextManager } = require('@opentelemetry/context-async-hooks');
const { trace, context, propagation } = require('@opentelemetry/api');
const { W3CTraceContextPropagator } = require('@opentelemetry/core');

const config = require('./config');

// Mirrors the setup in api-gateway (Go) and order-worker (Python): batch
// export spans over OTLP/gRPC to the shared collector, W3C trace context
// propagator so headers written by the other two services parse correctly.
function initTracer() {
  const provider = new NodeTracerProvider({
    resource: resourceFromAttributes({
      [ATTR_SERVICE_NAME]: config.serviceName,
    }),
    spanProcessors: [
      new BatchSpanProcessor(
        new OTLPTraceExporter({
          url: `${config.otelCollectorEndpoint.startsWith('http') ? '' : 'http://'}${config.otelCollectorEndpoint}`,
        })
      ),
    ],
  });

  const contextManager = new AsyncHooksContextManager();
  contextManager.enable();
  context.setGlobalContextManager(contextManager);

  provider.register();
  propagation.setGlobalPropagator(new W3CTraceContextPropagator());

  return trace.getTracer(config.serviceName);
}

module.exports = { initTracer };

File: services/order-worker/app/telemetry.py
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor


def init_tracer(service_name: str, collector_endpoint: str):
    """Sets up a TracerProvider exporting to the OTel Collector via OTLP/gRPC.
    Mirrors the Go setup in api-gateway so both services show up as siblings
    in the same trace in Jaeger."""
    resource = Resource.create({"service.name": service_name})
    provider = TracerProvider(resource=resource)

    exporter = OTLPSpanExporter(endpoint=collector_endpoint, insecure=True)
    provider.add_span_processor(BatchSpanProcessor(exporter))

    trace.set_tracer_provider(provider)
    return trace.get_tracer(service_name), provider